storage/filestore/dir.go

```diff
@@ -273,48 +265,61 @@ func (dir *Dir) Delete(ctx context.Context, ref storage.BlobRef) (err error) {
 	// _forwards_, this race should not occur because it is assumed that pieces are never
 	// rewritten with an _older_ storage format version.
 	for i := MinFormatVersionSupported; i <= MaxFormatVersionSupported; i++ {
-		verPath := blobPathForFormatVersion(pathBase, i)
-
-		// move to trash folder, this is allowed for some OS-es
-		moveErr = rename(verPath, trashPath)
-		if os.IsNotExist(moveErr) {
-			// no piece at that path; either it has a different storage format version or there
-			// was a concurrent delete. (this function is expected by callers to return a nil
-			// error in the case of concurrent deletes.)
-			continue
-		}
-		if moveErr != nil {
-			// piece could not be moved into the trash dir; we'll try removing it directly
-			trashPath = verPath
-		}
+		combinedErrors.Add(dir.DeleteWithStorageFormat(ctx, ref, i))
+	}
 
-		// try removing the file
-		err = os.Remove(trashPath)
+	return combinedErrors.Err()
+}
 
-		// ignore concurrent deletes
-		if os.IsNotExist(err) {
-			// something is happening at the same time as this; possibly a concurrent delete,
-			// or possibly a rewrite of the blob. keep checking for more versions.
-			continue
-		}
+// DeleteWithStorageFormat deletes the blob with the specified ref for one specific format version
+func (dir *Dir) DeleteWithStorageFormat(ctx context.Context, ref storage.BlobRef, formatVer storage.FormatVersion) (err error) {
+	defer mon.Task()(&ctx)(&err)
 
-		// the remove may have failed because of an open file handle. put it in a queue to be
-		// retried later.
-		if err != nil {
-			dir.mu.Lock()
-			dir.deleteQueue = append(dir.deleteQueue, trashPath)
-			dir.mu.Unlock()
-		}
+	pathBase, err := dir.blobToBasePath(ref)
+	if err != nil {
+		return err
+	}
+	trashPath := dir.blobToTrashPath(ref)
 
-		// ignore is-busy errors, they are still in the queue
-		// but no need to notify
-		if isBusy(err) {
-			err = nil
-		}
-		combinedErrors.Add(err)
+	verPath := blobPathForFormatVersion(pathBase, formatVer)
+
+	// move to trash folder, this is allowed for some OS-es
+	moveErr := rename(verPath, trashPath)
```

Comment:
Is it possible when a move is already in progress this line could execute twice? What's the consequences if that happens?

Reply:
I'm not sure what you mean by "executed twice." If concurrent deletes are occurring, this `rename` will error with "file not found," which is handled below.

Reply:
Do all filesystems make moves immediate? Can some take awhile? In that case a concurrent delete wouldn't cause a `file not found` in that case? I'm not sure if this is an issue or not.

Reply:
It is atomic, so a concurrent delete would result in a `file not found` for a move call. Either the move or the delete would win, so either the delete (and then file not found), or the move (and the other delete gets a file not found.)